### PR TITLE
Use exit callback instead of close callback [vim]

### LINKED ISSUE #307: Vim hang/unresponsive on BufEnter
I've noticed that when I open a file and save the changes with `:w` vim hangs and can't do anything with it until i press `Ctrl+C`. After some debugging (not entirely familiar with it, just changed the verbose to 12) it turns out that it stops at `kite#bufenter()` inside the `if !kite#utils#kite_running()` check. However I noticed that `kited` is already running but it still goes into the `if !kite#utils#kite_running()` part of the function. So my current hack is to disable the kited running check like:
```
function! kite#bufenter()
  if kite#languages#supported_by_plugin()
    call s:launch_kited()

"    if !kite#utils#kite_running()
"      call kite#status#status()
"      call s:start_status_timer()

"      call s:start_watching_for_kited()
"      return
"    endif

    call s:stop_watching_for_kited()

    if kite#languages#supported_by_kited()
      call s:disable_completion_plugins()
      call s:setup_options()
      call s:setup_events()
      call s:setup_mappings()
      call s:set_max_file_size()

      setlocal completefunc=kite#completion#complete

      call kite#events#event('focus')
      call kite#status#status()
      call s:start_status_timer()

      return
    end
  end

  " Buffer is not a supported language.
  call s:restore_options()
  call s:stop_status_timer()
endfunction
```
and so far it's working fine.

Other relevant informations:
my vim version:
```
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Aug 11 2020 17:00:59)
Included patches: 1-716
Modified by team+vim@tracker.debian.org
Compiled by team+vim@tracker.debian.org
Huge version with GTK3 GUI.  Features included (+) or not (-):
+acl               -farsi             +mouse_sgr         +tag_binary
+arabic            +file_in_path      -mouse_sysmouse    -tag_old_static
+autocmd           +find_in_path      +mouse_urxvt       -tag_any_white
+autochdir         +float             +mouse_xterm       +tcl
-autoservername    +folding           +multi_byte        +termguicolors
+balloon_eval      -footer            +multi_lang        +terminal
+balloon_eval_term +fork()            -mzscheme          +terminfo
+browse            +gettext           +netbeans_intg     +termresponse
++builtin_terms    -hangul_input      +num64             +textobjects
+byte_offset       +iconv             +packages          +textprop
+channel           +insert_expand     +path_extra        +timers
+cindent           +ipv6              +perl              +title
+clientserver      +job               +persistent_undo   +toolbar
+clipboard         +jumplist          +popupwin          +user_commands
+cmdline_compl     +keymap            +postscript        +vartabs
+cmdline_hist      +lambda            +printer           +vertsplit
+cmdline_info      +langmap           +profile           +virtualedit
+comments          +libcall           -python            +visual
+conceal           +linebreak         +python3           +visualextra
+cryptv            +lispindent        +quickfix          +viminfo
+cscope            +listcmds          +reltime           +vreplace
+cursorbind        +localmap          +rightleft         +wildignore
+cursorshape       +lua               -ruby              +wildmenu
+dialog_con_gui    +menu              +scrollbind        +windows
+diff              +mksession         +signs             +writebackup
+digraphs          +modify_fname      +smartindent       +X11
+dnd               +mouse             +sound             -xfontset
-ebcdic            +mouseshape        +spell             +xim
+emacs_tags        +mouse_dec         +startuptime       +xpm
+eval              +mouse_gpm         +statusline        +xsmp_interact
+ex_extra          -mouse_jsbterm     -sun_workshop      +xterm_clipboard
+extra_search      +mouse_netterm     +syntax            -xterm_save
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-3.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/gtk-3.0 -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -Wdate-time  -g -O2 -fdebug-prefix-map=/build/vim-It0QT5/vim-8.2.0716=. -fstack-protector-strong -Wformat -Werror=format-security -D_REENTRANT -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: gcc   -Wl,-E  -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -o vim   -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo  -lselinux  -lcanberra -lacl -lattr -lgpm -ldl  -L/usr/lib -llua5.2 -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/x86_64-linux-gnu/perl/5.30/CORE -lperl -ldl -lm -lpthread -lcrypt  -L/usr/lib/python3.8/config-3.8-x86_64-linux-gnu -lpython3.8 -lcrypt -lpthread -ldl -lutil -lm -lm -L/usr/lib/x86_64-linux-gnu -ltcl8.6 -ldl -lz -lpthread -lm
```

my .vimrc:
```
"vundle
set nocompatible
filetype on
filetype plugin on

set rtp+=~/.vim/bundle/Vundle.vim
call vundle#begin()

Plugin 'VundleVim/Vundle.vim'
"Git interface
Plugin 'tpope/vim-fugitive'
Plugin 'airblade/vim-gitgutter'
"Filesystem
Plugin 'scrooloose/nerdtree'
Plugin 'Xuyuanp/nerdtree-git-plugin'
Plugin 'jistr/vim-nerdtree-tabs'
Plugin 'kien/ctrlp.vim'

"search
Plugin 'google/vim-searchindex'
set rtp+=/usr/local/opt/fzf
Plugin 'junegunn/fzf.vim'
Plugin 'mileszs/ack.vim'
Plugin 'jremmen/vim-ripgrep'

"Python sytax checker
Plugin 'nvie/vim-flake8'
Plugin 'vim-scripts/indentpython.vim'

"Auto-completion
Plugin 'ervandew/supertab'
Plugin 'majutsushi/tagbar'
Plugin 'jiangmiao/auto-pairs'

"Colors
Plugin 'altercation/vim-colors-solarized'
Plugin 'jnurmine/Zenburn'

"Visual
Plugin 'vim-airline/vim-airline'
Plugin 'vim-airline/vim-airline-themes'
Plugin 'yuttie/comfortable-motion.vim'
Plugin 'nathanaelkane/vim-indent-guides'

"surround
Plugin 'tpope/vim-surround'

"utilities
Plugin 'christoomey/vim-tmux-navigator'

"Jenkinsfile"
Plugin 'martinda/jenkinsfile-vim-syntax'

"Julia
Plugin 'JuliaEditorSupport/julia-vim'

"Test
Plugin 'alfredodeza/pytest.vim'
Plugin 'benmills/vimux'

call vundle#end()

filetype plugin indent on
"let g:SimpylFold_docstring_preview = 1

"autocomplete
"let g:ycm_autoclose_preview_window_after_completion=1

"custom keys
let mapleader=" "
"map <leader>g  :YcmCompleter GoToDefinitionElseDeclaration<CR>
"
call togglebg#map("<F5>")
"colorscheme zenburn
"set guifont=Monaco:h14

let NERDTreeIgnore=['\.pyc$', '\~$'] "ignore files in NERDTree

"I don't like swap files
set noswapfile

"turn on numbering
set nu

"always show gutter
set signcolumn=yes

"Setting python filetype"
au! BufNewFile,BufReadPost *.{py,pyc,pyw} setlocal filetype=python foldmethod=indent foldlevelstart=20

"it would be nice to set tag files by the active virtualenv here
":set tags=~/mytags "tags for ctags and taglist
"omnicomplete
autocmd FileType python set omnifunc=pythoncomplete#Complete

" add yaml stuffs
au! BufNewFile,BufReadPost *.{yaml,yml} setlocal filetype=yaml foldmethod=indent
autocmd FileType yaml setlocal ts=2  sts=2  sw=2  expandtab

"------------Start Python PEP 8 stuff----------------
" Number of spaces that a pre-existing tab is equal to.
au BufRead, BufNewFile *py,*pyw,*.c,*.h set tabstop=4

"spaces for indents
au BufRead, BufNewFile *.py,*pyw set shiftwidth=4
au BufRead, BufNewFile *.py,*.pyw set expandtab
au BufRead, BufNewFile *.py set softtabstop=4

" Use the below highlight group when displaying bad whitespace is desired.
highlight BadWhitespace ctermbg=red guibg=red

" Display tabs at the beginning of a line in Python mode as bad.
au BufRead, BufNewFile *.py,*.pyw match BadWhitespace /^\t\+/
" Make trailing whitespace be flagged as bad.
au BufRead, BufNewFile *.py,*.pyw,*.c,*.h match BadWhitespace /\s\+$/

" Wrap text after a certain number of characters
au BufRead, BufNewFile *.py,*.pyw, setlocal textwidth=100

" Use UNIX (\n) line endings.
au BufNewFile *.py,*.pyw,*.c,*.h setlocal fileformat=unix

" Set the default file encoding to UTF-8:
set encoding=utf-8

" For full syntax highlighting:
let python_highlight_all=1
syntax on

" Keep indentation level from previous line:
autocmd FileType python set autoindent

" make backspaces more powerfull
set backspace=indent,eol,start

"Folding based on indentation:
autocmd FileType python set foldmethod=indent
"use space to open folds
"nnoremap <leader> za
"----------Stop python PEP 8 stuff--------------

"js stuff"
autocmd FileType javascript setlocal shiftwidth=2 tabstop=2

"NERDTree settings
let g:NERDTreeNodeDelimiter = "\u00a0"
autocmd StdinReadPre * let s:std_in=1
autocmd VimEnter * if argc() == 0 && !exists("s:std_in") | NERDTree | endif
autocmd VimEnter * if argc() == 1 && isdirectory(argv()[0]) && !exists("s:std_in") | exe 'NERDTree' argv()[0] | wincmd p | ene | endif
autocmd bufenter * if (winnr("$") == 1 && exists("b:NERDTree") && b:NERDTree.isTabTree()) | q | endif
map <C-n> :NERDTreeToggle<CR>

"Tagbar toggle
let g:Tlist_Ctags_Cmd='/usr/local/Cellar/ctags/5.8_1/bin/ctags'
nmap <C-t> :TagbarToggle<CR>

"Flake8 settings"
"autocmd BufWritePost *.py call Flake8()
autocmd FileType python map <buffer> <leader>f :call Flake8()<CR>

"set syntastic to passive on python
let g:syntastic_mode_map = {
        \ "mode": "active",
        \ "active_filetypes": [],
        \ "passive_filetypes": ["python"] }

"Kite settings
let g:kite_log=0
let g:kite_supported_languages = ['python', 'javascript', 'go']
let g:kite_auto_complete=1
let g:kite_tab_complete=1
set completeopt-=menu
set completeopt+=menuone   " show the popup menu even when there is only 1 match
set completeopt-=longest   " don't insert the longest common text
set completeopt+=preview   " don't show preview window
set completeopt+=noinsert  " don't insert any text until user chooses a match
set completeopt-=noselect  " select first match
autocmd CompleteDone * if !pumvisible() | pclose | endif

"customize airline
"let g:airline_section_y = "%{kite#statusline()}"
"let g:airline_section_y = '%{strftime("%c")}'

"kite docs shortcut
nmap <silent> <buffer> gK <Plug>(kite-docs)

"fix spacing function
function! Fix_spacing()
	%s/\<\(if\|for\|while\|switch\)(\s*/\1 (/ge     " if (
	%s/\((\) \(\S\)/\1\2/ge                         " no space after (
	%s/\n\n\n\+/\r\r/ge                             " single empty line
	%s/\([,;]\)\(\S\)/\1 \2/ge                      " space after, ;
	%s/\(\S\) \([,;)]\)/\1\2/ge                     " no space before, ;)
	%s/\s*$//ge                                     " no terminal space
endfunction

"Ack settings"
cnoreabbrev Ack Ack!
nnoremap <leader>a :Ack!<Space>

"Ripgrep
nnoremap <leader>r :Rg<Space>

"FZF
nnoremap <leader>b :Buffers<CR>

"Set toggle paste"
set pastetoggle=<F10>

"set mouse
set mouse=a

"resize pane shortcut
noremap <C-w>+ :resize +10<CR>
noremap <C-w>- :resize -10<CR>
noremap <C-w>< :vertical:resize -10<CR>
noremap <C-w>> :vertical:resize +10<CR>

"Test shortcuts
nmap <silent> t<C-f> <Esc>:Pytest file<CR>
nmap <silent> t<C-c> <Esc>:Pytest class<CR>
nmap <silent> t<C-m> <Esc>:Pytest method<CR>
nmap <silent> t<C-n> <Esc>:Pytest function<CR>
nmap <silent> t<C-p> <Esc>:Pytest project<CR>
nnoremap <leader>s :Pytest session<CR>

"Misc
set nowrap
highlight clear SignColumn 
```
Let me know if you need any other information. Thanks.

### LINKED ISSUE #313: Kite is conflicting with auto-pairs vim-closer and basically anything that uses the Enter key
I was using auto-pairs, however it stopped working properly after installing kite:

```javascript
const lala = () => {//cursor here}

// produces this:
const lala = () => {
//cursor here}

//instead of this
const lala = () => {
  //cursor here
}
```

Is there a work around for this?

### PR DESCRIPTION
**Pull request by airblade**
In certain circumstances the channel gets stuck in a closing state and is not actually closed until the user presses ctrl-c.  I don't know why.

Since we cannot rely on the close callback being called, use an exit callback instead.  Include a slight delay in the exit callback to allow time for any buffered data to hit the out callback.

Fixes #307.

### CODE REVIEW

**Comment by airblade** [comment]
I wasn't able to come up with a minimal way to reproduce the bug.  Here's the non-minimal way.

- Install [NERDTree](https://github.com/scrooloose/nerdtree) and Xuyuanp/nerdtree-git-plugin@a6f0c65df.
- Open a Python file which is tracked by git and has an unstaged changed.
- `:NERDTreeToggle` to open NERDTree.
- Expected: the NERDTree sidebar shows the Python file has a change and Vim remains responsive.
- Actual: Vim hangs until you press ctrl-c, at which point the expected behaviour resumes.

File: autoload/kite/async.vim [hunk 1]
```diff
@@ -93,12 +97,13 @@ function! s:on_stdout_vim(_channel, data) dict
   call add(self.stdoutbuffer, a:data)
 endfunction
 
-
-function! s:on_close_vim(channel) dict
+function! s:on_exit_vim(job, exit_status) dict
+  " Allow time for any buffered data to trigger out_cb.
+  " 5m is an educated guess.
+  sleep 5m
```
Comment (jansorg) [inline autoload/kite/async.vim:103]: is `5m` 5 milliseconds or something else?

**Review by jansorg — COMMENTED** [review]
Looks fine to me and it's working with my version of vim.
What would happen if the timeout in the exit handler is too short?

File: autoload/kite/async.vim (hunk 1 above)
Comment (airblade) [reply autoload/kite/async.vim:103]: 5 milliseconds.

It's possible for Vim to call the job's exit callback before all the data from stdout has been fed to the stdout callback.  I think this is something Vim should hide from the user and deal with, but instead the docs advise adding a "short" sleep at the top of the exit callback.

If the sleep is too short then it's possible to miss some data at the end from stdout.  I think 1ms would be enough time so 5ms really should be safe.

**Review by metalogical — APPROVED** [review]

**Comment by airblade** [comment]
Rebased and merged in c7d4a52.

**Pull request closed without merging**